### PR TITLE
Verify oe report with null format id

### PR DESCRIPTION
**Pull request by qiucwang**
Enable `oe_verify_evidence()` to specially handle legacy `oe_report` with a `NULL` format id.
So `oe_verify_evidence()` could process `oe_evidence` and legacy `oe_report` with same input parameters.

And adjust plugin testing so that all claims retrieved from legacy `oe_report` and raw `sgx_quote` will now be checked instead of just the `sgx_report_data` claim.

Signed-off-by: Qiucheng Wang <qiucwang@microsoft.com>

### CODE REVIEW

File: common/attest_plugin.c [hunk 1]
```diff
@@ -405,7 +427,7 @@ oe_result_t oe_verify_attestation_certificate_with_evidence(
     oe_cert_t cert = {0};
     uint8_t* report = NULL;
     size_t report_size = 0;
-    oe_report_header_t* header = NULL;
+    // oe_attestation_header_t* header = NULL;
```
Comment (dthaler) [inline common/attest_plugin.c:430]: delete commented out code

File: common/attest_plugin.c [hunk 2]
```diff
@@ -469,47 +491,11 @@ oe_result_t oe_verify_attestation_certificate_with_evidence(
...
-            report_size,
-            NULL,
-            0,
-            NULL,
-            0,
-            &claims,
-            &claims_length);
-    }
-
+    // report is either an oe_evidence which embed format id in its attestation
```
Comment (dthaler) [inline common/attest_plugin.c:494]:
```suggestion
    // A report is either an oe_evidence that embeds a format id in its attestation
```

File: tests/attestation_plugin/plugin/tests.c [hunk 3]
```diff
@@ -402,6 +402,175 @@ static void* _find_claim(
...
+    bool is_local,
+    const uint8_t* report_body,
+    const oe_uuid_t* format_id,
+    const oe_sgx_endorsements_t* sgx_endorsements,
+    const uint8_t* custom_claims_buffer,
+    size_t custom_claims_buffer_size)
+{
+    oe_report_t report;
+    const sgx_report_body_t* sgx_report_body;
+
```
Comment (dthaler) [inline tests/attestation_plugin/plugin/tests.c:418]:
```suggestion
```

File: tests/attestation_plugin/plugin/tests.c [hunk 4]
```diff
@@ -402,6 +402,175 @@ static void* _find_claim(
...
+    // Check SGX Report ISV Extended Product ID
+    value =
+        _find_claim(claims, claims_size, OE_CLAIM_SGX_ISV_EXTENDED_PRODUCT_ID);
+    OE_TEST(
+        value != NULL && memcmp(
+                             value,
+                             sgx_report_body->isvextprodid,
+                             sizeof(sgx_report_body->isvextprodid)) == 0);
+
+    // Check SGX Report Is Mode 64bit or Not
```
Comment (dthaler) [inline tests/attestation_plugin/plugin/tests.c:488]:
```suggestion
    // Check whether the SGX Report is Mode 64bit or not.
```

File: tests/attestation_plugin/plugin/tests.c [hunk 5]
```diff
@@ -402,6 +402,175 @@ static void* _find_claim(
...
+        {
+            value = claims[i].value;
+            OE_TEST(
+                value != NULL && memcmp(
+                                     value,
+                                     sgx_endorsements->items[j].data,
+                                     sgx_endorsements->items[j].size) == 0);
+        }
+    }
+    // Check custom claims / sgx_report_data.
```
Comment (dthaler) [inline tests/attestation_plugin/plugin/tests.c:558]:
```suggestion

    // Check custom claims / sgx_report_data.
```

**Review by dthaler — APPROVED** [review]
Looks good other than some editorial nits.

**Comment by qiucwang** [comment]
> Looks good other than some editorial nits.

File: common/attest_plugin.c (hunk 1 above)
Comment (qiucwang) [reply common/attest_plugin.c:430]: Thanks!

File: common/attest_plugin.c (hunk 2 above)
Comment (qiucwang) [reply common/attest_plugin.c:494]: Thanks!

File: common/attest_plugin.c [hunk 6]
```diff
@@ -301,61 +301,83 @@ oe_result_t oe_verify_evidence(
...
     if (!format_id)
     {
-        oe_attestation_header_t* evidence =
-            (oe_attestation_header_t*)evidence_buffer;
-
-        if (evidence_buffer_size < sizeof(oe_attestation_header_t) ||
-            evidence->version != OE_ATTESTATION_HEADER_VERSION)
+        // check whether evidence buffer structure is oe_report
+        oe_report_header_t* report = (oe_report_header_t*)evidence_buffer;
+        // oe_report_header_t size is smaller than oe_attestation_header_t size;
```
Comment (yentsanglee) [inline common/attest_plugin.c:306]: This assumption may not be true in the future (e.g. oe_attestation_header_t becomes smaller, even though highly unlikely).

To guarantee forward compatibility, I'd suggest combining the 2 if statements. I.e.

if (evidence_buffer_size >= sizeof(oe_report_header_t && 
    report->version == OE_REPORT_HEADER_VERSION)
    // treat as report

File: common/attest_plugin.c [hunk 7]
```diff
@@ -469,47 +491,11 @@ oe_result_t oe_verify_attestation_certificate_with_evidence(
     // find the extension
     OE_TRACE_VERBOSE("extract_x509_report_extension() succeeded");
 
-    if (oid_array_index >= 2) // oid_oe_evidence or oid_new_oe_evidence
```
Comment (yentsanglee) [inline common/attest_plugin.c:472]: This check is still needed even though you can use the same function/parameters to verify it.

```
If (oid_array_index >= 2)  // evidence format
    it's header version must be OE_ATTESTATION_HEADER_VERSION.
else
    it's header version must be OE_REPORT_HEADER_VERSION

```

**Comment by yentsanglee** [comment]
Looks good overall. Thanks for adding the extra test case. Only 2 comments.

File: common/attest_plugin.c (hunk 6 above)
Comment (qiucwang) [reply common/attest_plugin.c:306]: Make sense. Thanks for pointing out.

**Review by shnwc — COMMENTED** [review]
This is a change in public API. The design doc [Attestation_API_Proposal.md](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/Attestation_API_Proposal.md) and public header file [verifier.h](https://github.com/openenclave/openenclave/blob/master/include/openenclave/attestation/verifier.h) also need to be updated.

**Comment by qiucwang** [comment]
> This is a change in public API. The design doc [Attestation_API_Proposal.md](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/Attestation_API_Proposal.md) and public header file [verifier.h](https://github.com/openenclave/openenclave/blob/master/include/openenclave/attestation/verifier.h) also need to be updated.

Thanks for reminding. I'll update the design doc and public header.

**Comment by qiucwang** [comment]
bors try

**Review by shnwc — APPROVED** [review]

**Review by yentsanglee — APPROVED** [review]

**Comment by qiucwang** [comment]
bors retry

**Pull request merged**